### PR TITLE
Add | Allow use of platform gss for kerberos authentication

### PR DESCRIPTION
**Pull request by rene-ye**
Updated PR of #585 by @kilokahn due to merge conflicts

### CODE REVIEW

File: src/samples/platform-gss/pom.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.microsoft.sqlserver</groupId>
+            <artifactId>mssql-jdbc</artifactId>
+            <version>6.1.5.jre8-preview</version>
```
Comment (cheenamalhotra) [inline src/samples/platform-gss/pom.xml]: I think this should be updated to latest driver version (future release)

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java [hunk 2]
```diff
@@ -340,6 +340,7 @@ public String toString() {
     SEND_TIME_AS_DATETIME("sendTimeAsDatetime", true),
     TRANSPARENT_NETWORK_IP_RESOLUTION("TransparentNetworkIPResolution", true),
     TRUST_SERVER_CERTIFICATE("trustServerCertificate", false),
+    USE_PLATFORM_GSS_CREDENTIALS("usePlatformGssCredentials", false),
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java]: Let's also add this new connection property to SQLServerDataSource setter/getter API (also to interface) with Documentation in Javadocs

File: src/samples/platform-gss/main/java/PlatformGSS.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        System.setProperty("sun.security.jgss.native", "true");
+    }
+
+    public static void main(String... args) throws Exception {
+
+        Class.forName(DRIVER_CLASS_NAME).getConstructor().newInstance();
+
+        try (Connection con = createConnection()) {
+            System.out.println("Connection succesfully: " + con);
```
Comment (cheenamalhotra) [inline src/samples/platform-gss/main/java/PlatformGSS.java]: Please also add this to Connection Junit Test (with connection property and DataSource getConnection())

Comment (rene-ye) [reply src/samples/platform-gss/main/java/PlatformGSS.java]: This sample doesn't work on its own. It requires a keytab and some basic kerberos setup. I don't think we can add a junit test.

**Comment by akazakov** [comment]
This is really useful! Please merge.

**Comment by copumpkin** [comment]
Agreed, would love to see this merged. @cheenamalhotra it looks like @rene-ye addressed your comments, is there anything else blocking this?

**Comment by cheenamalhotra** [comment]
@akazakov @copumpkin 

We are targeting to review and take it forward to next preview release (**v7.3.0**).
Currently we're holding up for **v7.2.1** Hotfix Release, hence not merging any changes.

Appreciate your waiting, we will address this at earliest convenience.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataSource.java [hunk 4]
```diff
@@ -225,6 +225,19 @@ public GSSCredential getGSSCredentials() {
...
+    public void setUsePlatformGSSCredentials(boolean usePlatformGSS) {
+        setObjectProperty(connectionProps, SQLServerDriverBooleanProperty.USE_PLATFORM_GSS_CREDENTIALS.toString(),
+                usePlatformGSS);
+    }
+
+    @Override
+    public boolean getUsePlatformGSSCredentials() {
+        return getBooleanProperty(connectionProps,
+                SQLServerDriverBooleanProperty.USE_PLATFORM_GSS_CREDENTIALS.toString(),
+                SQLServerDriverBooleanProperty.ENCRYPT.getDefaultValue());
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataSource.java]: Change it to `SQLServerDriverBooleanProperty.USE_PLATFORM_GSS_CREDENTIALS.getDefaultValue()`.

**Comment by ulvii** [comment]
Hi @kilokahn, @akazakov, @copumpkin,
I was just checking if you considered using `gsscredential` connection property. You would need to fetch the default platform credential in your application and pass it to the driver: 
```
GSSCredential credential = GSSManager.getInstance().createCredential(null, GSSCredential.DEFAULT_LIFETIME, new Oid("1.2.840.113554.1.2.2"), GSSCredential.ACCEPT_ONLY);
SQLServerDataSource ds = new SQLServerDataSource();
ds.setGSSCredentials(credential);
```
Please also note that this property is only available through SQLServerDataSource (cannot be specified in connection string).

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 5]
```diff
@@ -1562,6 +1564,12 @@ Connection connectInternal(Properties propsIn,
                     ImpersonatedUserCred = (GSSCredential) activeConnectionProperties.get(sPropKey);
                     isUserCreatedCredential = true;
                 }
+                // There might be cases where we wish to use the platform GSS credentials
+                sPropKey = SQLServerDriverBooleanProperty.USE_PLATFORM_GSS_CREDENTIALS.toString();
+                sPropValue = activeConnectionProperties.getProperty(sPropKey);
+                if (sPropValue != null) {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: null != sPropValue to be consistent

**Comment by cheenamalhotra** [comment]
@kilokahn, @akazakov, @copumpkin,
Gentle reminder, we're waiting your response for @ulvii 's comment above.

**Comment by ulvii** [comment]
Closing the PR as the driver already provides this functionality via `gsscredential` connection property. Please let us know if you have any questions.

**Comment by fcorange** [comment]
@ulvii Tried to retrieve creds directly from GSSManager per your suggestion but looks like it has trouble locating Kerberos creds:

(I'm on Linux with Kerberos cache stored in keyring)

Sample code:

    val gssManager = GSSManager.getInstance()
    val krb5Mechanism = new Oid("1.2.840.113554.1.2.2")
    //val userName = gssManager.createName("<name>", GSSName.NT_USER_NAME)
    val credential = gssManager.createCredential(null, GSSCredential.DEFAULT_LIFETIME, krb5Mechanism, GSSCredential.ACCEPT_ONLY)

Error:

    Exception in thread "main" GSSException: No valid credentials provided (Mechanism level: Failed to find any Kerberos credentails)
	    at sun.security.jgss.krb5.Krb5AcceptCredential.getInstance(Krb5AcceptCredential.java:87)
	    at sun.security.jgss.krb5.Krb5MechFactory.getCredentialElement(Krb5MechFactory.java:129)
	    at sun.security.jgss.GSSManagerImpl.getCredentialElement(GSSManagerImpl.java:193)
	    at sun.security.jgss.GSSCredentialImpl.add(GSSCredentialImpl.java:427)

**Pull request closed without merging**